File: src/display/LedMatrix.ts
import * as gpio from 'rpio';
import { setInterval } from 'timers';

import Display from '../Display';
import Screen from '../Screen';

export enum MultiplexMode {
  ROW, COL,
}

// tslint:disable:no-bitwise

class LedMatrix implements Display {
  public static enabledListToByte(list: number[]): number {
    return list.reduce((data, enabled, i) => {
      if (enabled) {
        return data | (1 << i);
      } else {
        return data;
      }
    }, 0);
  }

  private mode: MultiplexMode = MultiplexMode.ROW;
  private readonly rows: number[];
  private readonly cols: number[];
  private do_show: boolean = false;

  constructor(rows: number[], cols: number[]) {
    this.cols = cols;
    this.rows = rows;

    this.openMatrix();
  }

  public tearDown() {
    this.closeMatrix();
  }

  public show(screen: Screen) {
    this.do_show = true;

    const matrix = screen.occupiedArea();

    return new Promise((resolve) => {
      while (this.do_show) {
        matrix.forEach((row, y) => {
          this.enableRow(y);

          row.forEach((is_set, x) => {
            if (is_set) {
              this.enableCol(x);
            }
          });

          this.disableRow(y);
        });
      }

      resolve();
    });
  }

  public clear(): void {
    this.do_show = false;
  }

  /**
   *
   * @param rows
   * @param cols
   */
  private openMatrix() {
    this.rows.concat(this.cols).forEach((pin) => {
      gpio.open(pin, gpio.OUTPUT, gpio.LOW);
    });
  }

  private closeMatrix() {
    this.rows.concat(this.cols).forEach((pin) => {
      gpio.close(pin);
    });
  }

  private enableRow(y: number): void {
    this.assertRowDefined(y);

    gpio.write(this.rows[y], gpio.HIGH);
  }

  private disableRow(y: number): void {
    this.assertRowDefined(y);

    gpio.write(this.rows[y], gpio.LOW);

    this.cols.forEach((col) => {
      gpio.write(col, gpio.LOW);
    });
  }

  private enableCol(x: number): void {
    this.assertColDefined(x);

    gpio.write(this.cols[x], gpio.HIGH);
  }

  private disableCol(x: number): void {
    this.assertColDefined(x);

    gpio.write(this.cols[x], gpio.LOW);
  }

  private assertColDefined(x: number): void {
    if (this.cols[x] === undefined) {
      throw new Error(`col ${x} not defined`);
    }
  }

  private assertRowDefined(y: number): void {
    if (this.rows[y] === undefined) {
      throw new Error(`row ${y} not defined`);
    }
  }
}

export default LedMatrix;

File: src/display/LedMatrix.py
import RPi.GPIO as GPIO

from src.Display import Display
from src.Screen import Screen
from src.StoppableThread import StoppableThread

def pinOn(pin):
  GPIO.output(pin, GPIO.HIGH)

def pinOff(pin):
  GPIO.output(pin, GPIO.LOW)

def displayMatrix(matrix, rows, cols):
  for i, row in enumerate(matrix):
    pinOn(rows[i])

    for j, is_on in enumerate(row):
      if is_on:
        pinOn(cols[j])
        pinOff(cols[j])

    pinOff(rows[i])

def _show_worker(display, matrix):
  displayMatrix(matrix, display.rows, display.cols)

class LedMatrix(Display):
  @staticmethod
  def withPins(rows, cols, mode):
    GPIO.setmode(mode)
    return LedMatrix(rows, cols)

  def __init__(self, rows, cols):
    self.rows = rows
    self.cols = cols

    self._openMatrix()
    self._worker = None

  def tearDown(self):
    self.clear()
    self._closeMatrix()

  def show(self, screen):
    self.clear()

    self._worker = self._setup_worker(screen)
    self._worker.start()

  def debug(self, text):
    print(text)

  def clear(self):
    if self._worker is not None:
      self._worker.stop(blocking=True)

    self.reset()

  @property
  def _pins(self):
    return self.rows + self.cols

  def _openMatrix(self):
    GPIO.setmode(GPIO.BOARD)
    for pin in self._pins:
      GPIO.setup(pin, GPIO.OUT)
      
  def _closeMatrix(self):
    GPIO.cleanup(self._pins)

  def reset(self):
    for pin in self._pins:
      pinOff(pin)

  def enableRow(self, y):
    pinOn(self.rows[y])

  def disableRow(self, y):
    pinOff(self.rows[y])

    for pin in self.cols:
      pinOff(pin) 

  def enableCol(self, x):
    pinOn(self.cols[x])

  def disableCol(self, x):
    pinOff(self.cols[x])

  def _setup_worker(self, screen):
    return StoppableThread(
      target=_show_worker,
      kwargs={
        'matrix': screen.occupiedArea(),
        'display': self
      }
    )
